app_builder/mix.exs

```diff
@@ -13,7 +13,7 @@ defmodule AppBuilder.MixProject do
 
   def application do
     [
-      extra_applications: [:logger, :eex, :inets, :ssl, :crypto]
+      extra_applications: [:logger, :eex, :inets, :ssl, :crypto, :wx]
```

Comment:
I don't think this should be here. We will allow people to use AppBuilder without wx after all, right? I think everything wx related should be on a AppBuilder.Wx module and we need to tell people to add wx as a dependency. Otherwise we will require wx inside Docker and what not.

Reply:
If we really want to go this route, then we need to pass `runtime: false` to the `:app_builder` dep.

Reply:
I though that in the Windows launcher script it's better to have:

```
bin/app rpc "AppBuilder.Windows.send_wx_events(MyApp, input)"
```

than:

```
bin/app rpc "MyApp.rpc(input)"
```

That is, with the former it works out of the box because the user is already implementing app event handlers (`def handle_info({:open_url, ...` & `def handle_info({:open_file, ...`) because they need it for Mac. In the latter, we ask the user to implement a function which would look like this:

```elixir
# app got activated
def rpc("") do
end

def rpc("open_url:" <> url) do
end

def rpc("open_path:" <> path) do
end
```

And so need to document that they need to define a function called `rpc` (terrible name but can't think of a better one) with the following possible inputs.

But I guess I'm not documenting anywhere that they need to do `handle_info({:open_url, ...` either.


I think the proper way would be to establish a contract:

```elixir
defmodule AppBuilder.App do
  @callback reopen_app(state) :: :ok
  @callback open_url(url) :: :ok
  @callback open_file(path) :: :ok
end

defmodule LivebookApp do
  use GenServer
  @behaviour AppBuilder.App

  @impl true
  def open_url(url) do
    send(__MODULE__, {:open_url, url})
  end

  @impl true
  def handle_info({:open_url, url}, state) do
    open_browser(Livebook.import_url(url))
    {:noreply, state}
  end
end
```

or even:

```elixir
defmodule LivebookApp do
  use GenServer

  use AppBuilder.App
  # adds:
  #
  #   def __open_url__(url) do
  #     send(__MODULE__, {:open_url, url})
  #   end
  #
  # which the launcher will call.

  @impl true
  def open_url(url, state) do
    open_browser(Livebook.import_url(url))
    {:noreply, state}
  end
end
```

I guess it actually looks pretty good if I may say so but I really don't know how far I should go with this at this point.

Reply:
Done in f6db279e4eeb05bfe1b2100e49d4338fe34632f5, thanks!